// @viniciustrindade/gdownjs/dist/cache.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateFileHash = calculateFileHash;
exports.loadMetadata = loadMetadata;
exports.saveMetadata = saveMetadata;
exports.verifyFileHash = verifyFileHash;
exports.isFileValid = isFileValid;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const crypto = __importStar(require("crypto"));
const CACHE_DIR = path.join(process.env.HOME || process.env.USERPROFILE || ".", ".gdown_cache");
/**
 * Ensure cache directory exists
 */
function ensureCacheDir() {
    if (!fs.existsSync(CACHE_DIR)) {
        fs.mkdirSync(CACHE_DIR, { recursive: true });
    }
}
/**
 * Get cache file path for a file ID
 */
function getCacheFilePath(fileId) {
    ensureCacheDir();
    return path.join(CACHE_DIR, `${fileId}.json`);
}
/**
 * Calculate file hash (SHA256)
 */
function calculateFileHash(filePath) {
    return new Promise((resolve, reject) => {
        const hash = crypto.createHash("sha256");
        const stream = fs.createReadStream(filePath);
        stream.on("data", (data) => hash.update(data));
        stream.on("end", () => resolve(hash.digest("hex")));
        stream.on("error", reject);
    });
}
/**
 * Load metadata from cache
 */
function loadMetadata(fileId) {
    try {
        const cachePath = getCacheFilePath(fileId);
        if (fs.existsSync(cachePath)) {
            const content = fs.readFileSync(cachePath, "utf8");
            const entry = JSON.parse(content);
            return {
                id: entry.id,
                name: entry.name,
                size: entry.size,
                hash: entry.hash,
                mimeType: entry.mimeType,
            };
        }
    }
    catch (_) {
        // Ignore cache errors
    }
    return null;
}
/**
 * Save metadata to cache
 */
function saveMetadata(metadata) {
    try {
        const cachePath = getCacheFilePath(metadata.id);
        const entry = {
            id: metadata.id,
            name: metadata.name,
            size: metadata.size,
            hash: metadata.hash,
            mimeType: metadata.mimeType,
            timestamp: Date.now(),
        };
        fs.writeFileSync(cachePath, JSON.stringify(entry, null, 2), "utf8");
    }
    catch (_) {
        // Ignore cache errors
    }
}
/**
 * Verify file hash matches cached hash
 */
async function verifyFileHash(filePath, expectedHash) {
    try {
        const actualHash = await calculateFileHash(filePath);
        return actualHash === expectedHash;
    }
    catch (_) {
        return false;
    }
}
/**
 * Check if file exists and hash matches (for --verify option)
 */
async function isFileValid(filePath, expectedHash) {
    if (!fs.existsSync(filePath)) {
        return false;
    }
    if (expectedHash) {
        return await verifyFileHash(filePath, expectedHash);
    }
    return true;
}
