// @viniciustrindade/gdownjs/dist/download.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadFile = downloadFile;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const https = __importStar(require("https"));
const http = __importStar(require("http"));
const url_1 = require("url");
const errors_1 = require("./errors");
const parser_1 = require("./parser");
const cookies_1 = require("./cookies");
const cache_1 = require("./cache");
const formats_1 = require("./formats");
const utils_1 = require("./utils");
/**
 * Extract confirmation token from HTML response
 */
function extractConfirmToken(html) {
    const tokenMatch = html.match(/confirm=([0-9A-Za-z_]+)&amp;id=/);
    if (tokenMatch) {
        return tokenMatch[1];
    }
    const inputMatch = html.match(/name="confirm"\s+value="([0-9A-Za-z_]+)"/);
    if (inputMatch) {
        return inputMatch[1];
    }
    return null;
}
/**
 * Perform a download request
 */
async function performDownloadRequest(urlStr, destDir, preferredName, jar, options, onProgress) {
    return new Promise((resolve, reject) => {
        const urlObj = new url_1.URL(urlStr);
        const client = urlObj.protocol === "https:" ? https : http;
        const headers = {
            "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36",
            Accept: "*/*",
        };
        const cookie = (0, cookies_1.cookieHeader)(jar);
        if (cookie) {
            headers.Cookie = cookie;
        }
        const requestOptions = {
            headers,
            rejectUnauthorized: !options.noCheckCertificate,
        };
        const req = client.get(urlStr, requestOptions, (res) => {
            (0, cookies_1.updateCookiesFromHeaders)(jar, res.headers["set-cookie"]);
            // Handle rate limiting
            if (res.statusCode === 429) {
                res.resume();
                resolve({ success: false, body: "Rate limited (429)", statusCode: 429 });
                return;
            }
            // Handle redirects
            if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
                res.resume();
                const nextUrl = new url_1.URL(res.headers.location, urlStr).toString();
                performDownloadRequest(nextUrl, destDir, preferredName, jar, options, onProgress)
                    .then(resolve)
                    .catch(reject);
                return;
            }
            // Handle successful download
            const disposition = res.headers["content-disposition"];
            if (disposition && res.statusCode === 200) {
                const inferredName = (0, utils_1.extractFilenameFromDisposition)(disposition);
                const fallback = (0, utils_1.sanitizeFileName)(preferredName || "download", "download");
                const baseName = (0, utils_1.sanitizeFileName)(inferredName || fallback, fallback);
                const finalName = (0, utils_1.ensureUniqueName)(destDir, baseName);
                const filePath = path.join(destDir, finalName);
                const totalBytes = parseInt(res.headers["content-length"] || "0", 10);
                let downloadedBytes = 0;
                const fileStream = fs.createWriteStream(filePath);
                res.on("data", (chunk) => {
                    downloadedBytes += chunk.length;
                    if (onProgress) {
                        onProgress({
                            bytesDownloaded: downloadedBytes,
                            totalBytes: totalBytes > 0 ? totalBytes : undefined,
                            percentage: totalBytes > 0 ? (downloadedBytes / totalBytes) * 100 : undefined,
                        });
                    }
                });
                res.pipe(fileStream);
                fileStream.on("finish", () => {
                    resolve({ success: true, filePath, statusCode: res.statusCode });
                });
                fileStream.on("error", reject);
            }
            else {
                // Read response body for error or confirmation token
                const chunks = [];
                res.on("data", (chunk) => chunks.push(chunk));
                res.on("end", () => {
                    try {
                        const buffer = Buffer.concat(chunks);
                        const body = (0, utils_1.decodeBody)(buffer, res.headers["content-encoding"]);
                        resolve({ success: false, body, statusCode: res.statusCode });
                    }
                    catch (err) {
                        reject(err);
                    }
                });
            }
        });
        req.on("error", reject);
        req.setTimeout(60000, () => {
            req.destroy();
            reject(new Error("Request timeout"));
        });
    });
}
/**
 * Download a file from Google Drive
 */
async function downloadFile(urlOrId, output, options = {}) {
    const { format, verify = false, quiet = false, verbose = false, noCookies = false, continue: resume = false, } = options;
    // Parse URL or use direct ID
    let fileId;
    let resourceKey;
    let fileType;
    if (options.id) {
        fileId = options.id;
        resourceKey = options.resourceKey;
    }
    else {
        const parsed = (0, parser_1.parseGoogleDriveUrl)(urlOrId);
        if (!parsed) {
            throw new Error(`Unable to extract Google Drive ID from: ${urlOrId}`);
        }
        fileId = parsed.id;
        resourceKey = parsed.resourceKey || options.resourceKey;
        fileType = parsed.type;
    }
    // Determine output path
    let outputPath;
    if (output) {
        outputPath = path.isAbsolute(output) ? output : path.resolve(process.cwd(), output);
    }
    else {
        outputPath = path.resolve(process.cwd(), fileId);
    }
    // Check if directory or file
    const isDirectory = fs.existsSync(outputPath) && fs.statSync(outputPath).isDirectory();
    const destDir = isDirectory ? outputPath : path.dirname(outputPath);
    const preferredName = isDirectory ? undefined : path.basename(outputPath);
    // Ensure destination directory exists
    if (!fs.existsSync(destDir)) {
        fs.mkdirSync(destDir, { recursive: true });
    }
    // Load cookies if not disabled
    const jar = noCookies ? {} : (0, cookies_1.loadCookies)();
    // Check cache if verify is enabled
    if (verify) {
        const metadata = (0, cache_1.loadMetadata)(fileId);
        if (metadata && metadata.hash) {
            const finalPath = preferredName
                ? path.join(destDir, preferredName)
                : path.join(destDir, metadata.name || fileId);
            if (await (0, cache_1.isFileValid)(finalPath, metadata.hash)) {
                if (!quiet) {
                    console.log(`File already exists and hash matches: ${finalPath}`);
                }
                return finalPath;
            }
        }
    }
    // Check for resume
    let finalOutputPath = preferredName
        ? path.join(destDir, preferredName)
        : path.join(destDir, fileId);
    if (resume && fs.existsSync(finalOutputPath)) {
        if (!quiet) {
            console.log(`Resuming download: ${finalOutputPath}`);
        }
        // Note: Full resume implementation would require Range headers
        // For now, we'll just continue with normal download
    }
    // Build download URL
    let downloadUrl = (0, parser_1.buildDownloadUrl)(fileId, resourceKey);
    if (format && fileType) {
        const exportFormat = (0, formats_1.getExportFormat)(fileType, format);
        if (exportFormat) {
            downloadUrl = (0, parser_1.buildDownloadUrl)(fileId, resourceKey, exportFormat);
        }
    }
    // Progress callback
    const onProgress = quiet
        ? undefined
        : (progress) => {
            if (verbose) {
                const total = progress.totalBytes
                    ? ` / ${(progress.totalBytes / (1024 * 1024)).toFixed(2)} MB`
                    : "";
                const pct = progress.percentage ? ` (${progress.percentage.toFixed(1)}%)` : "";
                process.stdout.write(`\rDownloading: ${(progress.bytesDownloaded / (1024 * 1024)).toFixed(2)} MB${total}${pct}`);
            }
        };
    // Download with confirmation token handling
    let confirmToken = null;
    const maxAttempts = 15;
    for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
        const url = new url_1.URL(downloadUrl);
        if (confirmToken) {
            url.searchParams.set("confirm", confirmToken);
        }
        try {
            const result = await performDownloadRequest(url.toString(), destDir, preferredName, jar, options, onProgress);
            if (result.success && result.filePath) {
                if (!quiet && verbose) {
                    process.stdout.write("\n");
                }
                if (!quiet) {
                    console.log(`Downloaded: ${result.filePath}`);
                }
                // Save cookies
                if (!noCookies) {
                    (0, cookies_1.saveCookies)(jar);
                }
                // Calculate and save hash if verify is enabled
                if (verify) {
                    const hash = await (0, cache_1.calculateFileHash)(result.filePath);
                    (0, cache_1.saveMetadata)({
                        id: fileId,
                        name: path.basename(result.filePath),
                        size: fs.statSync(result.filePath).size,
                        hash,
                    });
                }
                return result.filePath;
            }
            const body = result.body || "";
            // Check for rate limiting
            if (result.statusCode === 429 ||
                body.includes("rate limit") ||
                body.includes("quota") ||
                body.includes("429") ||
                body.includes("Too many requests")) {
                const backoffDelay = Math.min(5000 * Math.pow(2, attempt), 120000);
                if (!quiet) {
                    console.log(`Rate limited, waiting ${(backoffDelay / 1000).toFixed(1)}s before retry (attempt ${attempt + 1}/${maxAttempts})...`);
                }
                await (0, utils_1.sleep)(backoffDelay);
                continue;
            }
            // Check for permission errors
            if (body.includes("permission") || body.includes("access denied") || result.statusCode === 403) {
                throw new errors_1.PermissionError("Permission denied. The file may not be publicly accessible.");
            }
            // Check for file not found
            if (body.includes("not found") || result.statusCode === 404) {
                throw new errors_1.FileNotFoundError("File not found on Google Drive.");
            }
            // Extract confirmation token
            const token = extractConfirmToken(body);
            if (!token) {
                if (attempt < maxAttempts - 1) {
                    const backoffDelay = Math.min(3000 * Math.pow(1.5, attempt), 30000);
                    if (!quiet) {
                        console.log(`Unable to obtain token, waiting ${(backoffDelay / 1000).toFixed(1)}s before retry (attempt ${attempt + 1}/${maxAttempts})...`);
                    }
                    await (0, utils_1.sleep)(backoffDelay);
                    continue;
                }
                throw new Error("Unable to obtain Google Drive confirmation token");
            }
            confirmToken = token;
            if (!quiet) {
                console.log(`Confirmation required, retrying download (attempt ${attempt + 1}/${maxAttempts})...`);
            }
            if (attempt < maxAttempts - 1) {
                const delay = Math.min(2000 + attempt * 1000, 15000);
                await (0, utils_1.sleep)(delay);
            }
        }
        catch (err) {
            const errorMsg = err instanceof Error ? err.message : String(err);
            if (errorMsg.includes("429") ||
                errorMsg.includes("rate limit") ||
                errorMsg.includes("quota") ||
                errorMsg.includes("Too many requests")) {
                const backoffDelay = Math.min(5000 * Math.pow(2, attempt), 120000);
                if (!quiet) {
                    console.log(`Rate limit error, waiting ${(backoffDelay / 1000).toFixed(1)}s before retry (attempt ${attempt + 1}/${maxAttempts})...`);
                }
                await (0, utils_1.sleep)(backoffDelay);
                continue;
            }
            if (attempt === maxAttempts - 1) {
                throw err;
            }
            const backoffDelay = Math.min(2000 * Math.pow(1.5, attempt), 20000);
            if (!quiet) {
                console.log(`Error: ${errorMsg}, waiting ${(backoffDelay / 1000).toFixed(1)}s before retry (attempt ${attempt + 1}/${maxAttempts})...`);
            }
            await (0, utils_1.sleep)(backoffDelay);
        }
    }
    throw new Error("Exceeded maximum attempts while downloading from Google Drive");
}
