// @viniciustrindade/gdownjs/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationError = exports.PermissionError = exports.FileNotFoundError = exports.RateLimitError = exports.GdownError = void 0;
class GdownError extends Error {
    constructor(message) {
        super(message);
        this.name = "GdownError";
        Object.setPrototypeOf(this, GdownError.prototype);
    }
}
exports.GdownError = GdownError;
class RateLimitError extends GdownError {
    constructor(message = "Rate limit exceeded") {
        super(message);
        this.name = "RateLimitError";
        Object.setPrototypeOf(this, RateLimitError.prototype);
    }
}
exports.RateLimitError = RateLimitError;
class FileNotFoundError extends GdownError {
    constructor(message = "File not found") {
        super(message);
        this.name = "FileNotFoundError";
        Object.setPrototypeOf(this, FileNotFoundError.prototype);
    }
}
exports.FileNotFoundError = FileNotFoundError;
class PermissionError extends GdownError {
    constructor(message = "Permission denied") {
        super(message);
        this.name = "PermissionError";
        Object.setPrototypeOf(this, PermissionError.prototype);
    }
}
exports.PermissionError = PermissionError;
class VerificationError extends GdownError {
    constructor(message = "File verification failed") {
        super(message);
        this.name = "VerificationError";
        Object.setPrototypeOf(this, VerificationError.prototype);
    }
}
exports.VerificationError = VerificationError;
