// @viniciustrindade/gdownjs/dist/formats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PRESENTATION_FORMATS = exports.SPREADSHEET_FORMATS = exports.DOCUMENT_FORMATS = void 0;
exports.getDocumentFormat = getDocumentFormat;
exports.getSpreadsheetFormat = getSpreadsheetFormat;
exports.getPresentationFormat = getPresentationFormat;
exports.getExportFormat = getExportFormat;
exports.getFormatMimeType = getFormatMimeType;
/**
 * Format mappings for Google Docs/Sheets/Slides
 */
exports.DOCUMENT_FORMATS = {
    pdf: "pdf",
    docx: "docx",
    odt: "odt",
    rtf: "rtf",
    txt: "txt",
    html: "html",
    epub: "epub",
};
exports.SPREADSHEET_FORMATS = {
    xlsx: "xlsx",
    ods: "ods",
    csv: "csv",
    tsv: "tsv",
    pdf: "pdf",
    html: "html",
};
exports.PRESENTATION_FORMATS = {
    pptx: "pptx",
    odp: "odp",
    pdf: "pdf",
    txt: "txt",
    png: "png",
    jpg: "jpg",
    svg: "svg",
};
/**
 * Get export format for Google Docs
 */
function getDocumentFormat(format) {
    const normalized = format.toLowerCase();
    return exports.DOCUMENT_FORMATS[normalized] || null;
}
/**
 * Get export format for Google Sheets
 */
function getSpreadsheetFormat(format) {
    const normalized = format.toLowerCase();
    return exports.SPREADSHEET_FORMATS[normalized] || null;
}
/**
 * Get export format for Google Slides
 */
function getPresentationFormat(format) {
    const normalized = format.toLowerCase();
    return exports.PRESENTATION_FORMATS[normalized] || null;
}
/**
 * Get export format based on file type
 */
function getExportFormat(fileType, format) {
    switch (fileType) {
        case "document":
            return getDocumentFormat(format);
        case "spreadsheet":
            return getSpreadsheetFormat(format);
        case "presentation":
            return getPresentationFormat(format);
        default:
            return null;
    }
}
/**
 * Get MIME type for export format
 */
function getFormatMimeType(format) {
    const mimeTypes = {
        pdf: "application/pdf",
        docx: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        odt: "application/vnd.oasis.opendocument.text",
        rtf: "application/rtf",
        txt: "text/plain",
        html: "text/html",
        epub: "application/epub+zip",
        xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        ods: "application/vnd.oasis.opendocument.spreadsheet",
        csv: "text/csv",
        tsv: "text/tab-separated-values",
        pptx: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        odp: "application/vnd.oasis.opendocument.presentation",
        png: "image/png",
        jpg: "image/jpeg",
        svg: "image/svg+xml",
    };
    return mimeTypes[format.toLowerCase()] || "application/octet-stream";
}
