// @viniciustrindade/gdownjs/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationError = exports.PermissionError = exports.FileNotFoundError = exports.RateLimitError = exports.GdownError = exports.downloadFolder = exports.downloadFile = void 0;
exports.download = download;
const download_1 = require("./download");
const folder_1 = require("./folder");
/**
 * Download a file or folder from Google Drive
 *
 * @param url - Google Drive URL or file/folder ID
 * @param output - Output path (file or directory)
 * @param options - Download options
 * @returns Path to downloaded file or folder
 */
async function download(url, output, options = {}) {
    if (options.folder) {
        return (0, folder_1.downloadFolder)(url, output, options);
    }
    return (0, download_1.downloadFile)(url, output, options);
}
/**
 * Download a file from Google Drive
 *
 * @param url - Google Drive URL or file ID
 * @param output - Output file path
 * @param options - Download options
 * @returns Path to downloaded file
 */
var download_2 = require("./download");
Object.defineProperty(exports, "downloadFile", { enumerable: true, get: function () { return download_2.downloadFile; } });
/**
 * Download a folder from Google Drive
 *
 * @param url - Google Drive folder URL or folder ID
 * @param output - Output directory path
 * @param options - Download options
 * @returns Path to downloaded folder
 */
var folder_2 = require("./folder");
Object.defineProperty(exports, "downloadFolder", { enumerable: true, get: function () { return folder_2.downloadFolder; } });
/**
 * Custom error classes
 */
var errors_1 = require("./errors");
Object.defineProperty(exports, "GdownError", { enumerable: true, get: function () { return errors_1.GdownError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return errors_1.RateLimitError; } });
Object.defineProperty(exports, "FileNotFoundError", { enumerable: true, get: function () { return errors_1.FileNotFoundError; } });
Object.defineProperty(exports, "PermissionError", { enumerable: true, get: function () { return errors_1.PermissionError; } });
Object.defineProperty(exports, "VerificationError", { enumerable: true, get: function () { return errors_1.VerificationError; } });
