// @viniciustrindade/gdownjs/dist/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeBody = decodeBody;
exports.fetchText = fetchText;
exports.extractFilenameFromDisposition = extractFilenameFromDisposition;
exports.sanitizeFileName = sanitizeFileName;
exports.ensureUniqueName = ensureUniqueName;
exports.sleep = sleep;
exports.formatBytes = formatBytes;
const https = __importStar(require("https"));
const http = __importStar(require("http"));
const zlib = __importStar(require("zlib"));
const errors_1 = require("./errors");
const cookies_1 = require("./cookies");
/**
 * Decode response body based on content encoding
 */
function decodeBody(buffer, encoding) {
    if (encoding === "gzip") {
        return zlib.gunzipSync(buffer).toString("utf8");
    }
    if (encoding === "deflate") {
        return zlib.inflateSync(buffer).toString("utf8");
    }
    return buffer.toString("utf8");
}
/**
 * Fetch text from URL with retries and rate limit handling
 */
async function fetchText(url, options = {}) {
    const { jar = {}, maxRetries = 5, retryCount = 0, proxy, noCheckCertificate = false, } = options;
    return new Promise((resolve, reject) => {
        const urlObj = new URL(url);
        const client = urlObj.protocol === "https:" ? https : http;
        const requestOptions = {
            headers: {
                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36",
                Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            },
            rejectUnauthorized: !noCheckCertificate,
        };
        const cookie = (0, cookies_1.cookieHeader)(jar);
        if (cookie) {
            requestOptions.headers = {
                ...requestOptions.headers,
                Cookie: cookie,
            };
        }
        if (proxy) {
            // Proxy support would require additional implementation
            // For now, we'll skip proxy in fetchText
        }
        const req = client.get(url, requestOptions, (res) => {
            // Handle redirects
            if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
                res.resume();
                const nextUrl = new URL(res.headers.location, url).toString();
                fetchText(nextUrl, { ...options, retryCount }).then(resolve).catch(reject);
                return;
            }
            // Handle rate limiting
            if (res.statusCode === 429) {
                res.resume();
                if (retryCount < maxRetries) {
                    const backoffDelay = Math.min(2000 * Math.pow(2, retryCount), 60000);
                    setTimeout(() => {
                        fetchText(url, { ...options, retryCount: retryCount + 1 }).then(resolve).catch(reject);
                    }, backoffDelay);
                    return;
                }
                reject(new errors_1.RateLimitError(`Rate limited after ${maxRetries} retries`));
                return;
            }
            // Handle errors
            if (res.statusCode && res.statusCode >= 400) {
                res.resume();
                reject(new Error(`HTTP ${res.statusCode} when fetching ${url}`));
                return;
            }
            // Read response
            (0, cookies_1.updateCookiesFromHeaders)(jar, res.headers["set-cookie"]);
            const chunks = [];
            res.on("data", (chunk) => chunks.push(chunk));
            res.on("end", () => {
                try {
                    const buffer = Buffer.concat(chunks);
                    const text = decodeBody(buffer, res.headers["content-encoding"]);
                    resolve(text);
                }
                catch (err) {
                    reject(err);
                }
            });
        });
        req.on("error", async (err) => {
            if (retryCount < maxRetries) {
                const backoffDelay = Math.min(2000 * Math.pow(2, retryCount), 30000);
                await new Promise((r) => setTimeout(r, backoffDelay));
                fetchText(url, { ...options, retryCount: retryCount + 1 }).then(resolve).catch(reject);
            }
            else {
                reject(err);
            }
        });
        req.setTimeout(60000, () => {
            req.destroy();
            reject(new Error("Request timeout"));
        });
    });
}
/**
 * Extract filename from Content-Disposition header
 */
function extractFilenameFromDisposition(disposition) {
    // Try UTF-8 encoded filename first
    const utfMatch = disposition.match(/filename\*=UTF-8''([^;]+)/i);
    if (utfMatch) {
        try {
            return decodeURIComponent(utfMatch[1]);
        }
        catch (_) {
            // Fall through to regular filename
        }
    }
    // Try regular filename
    const match = disposition.match(/filename="([^"]+)"/i);
    if (match) {
        return match[1];
    }
    // Try unquoted filename
    const unquotedMatch = disposition.match(/filename=([^;]+)/i);
    if (unquotedMatch) {
        return unquotedMatch[1].trim();
    }
    return null;
}
/**
 * Sanitize filename for filesystem
 */
function sanitizeFileName(name, fallback) {
    const cleaned = (name || "")
        .replace(/[<>:"/\\|?*\x00-\x1F]/g, "_")
        .trim();
    if (cleaned.length === 0) {
        return fallback;
    }
    return cleaned;
}
/**
 * Ensure unique filename in directory
 */
function ensureUniqueName(dir, desiredName) {
    const path = require("path");
    const fs = require("fs");
    const ext = path.extname(desiredName);
    const base = path.basename(desiredName, ext);
    let candidate = desiredName;
    let counter = 1;
    while (fs.existsSync(path.join(dir, candidate))) {
        candidate = `${base} (${counter})${ext}`;
        counter += 1;
    }
    return candidate;
}
/**
 * Sleep for specified milliseconds
 */
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
/**
 * Format bytes to human readable string
 */
function formatBytes(bytes) {
    if (bytes === 0)
        return "0 B";
    const k = 1024;
    const sizes = ["B", "KB", "MB", "GB", "TB"];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return `${(bytes / Math.pow(k, i)).toFixed(2)} ${sizes[i]}`;
}
